File: src/main/kotlin/xyz/mahmoudahmed/dsl/TypedTargetPathBuilder.java
package xyz.mahmoudahmed.dsl;

import java.util.function.BiConsumer;
import java.util.function.Function;
import java.util.function.Predicate;

public class TypedTargetPathBuilder<S, T, V, U> {
    private final TypedTargetBuilder<S, T> parent;
    private final Function<S, V> sourceAccessor;
    private final BiConsumer<T, U> targetSetter;
    private Function<V, Object> transformation;
    private Predicate<S> condition;

    TypedTargetPathBuilder(TypedTargetBuilder<S, T> parent, Function<S, V> sourceAccessor,
                           BiConsumer<T, U> targetSetter, Function<V, Object> transformation,
                           Predicate<S> condition) {
        this.parent = parent;
        this.sourceAccessor = sourceAccessor;
        this.targetSetter = targetSetter;
        this.transformation = transformation;
        this.condition = condition;
    }

    public TypedTargetBuilder<S, T> using(Function<V, U> function) {
        this.transformation = (Function<V, Object>)(Function<?, ?>) function;
        finalizeMappingRule();
        return parent;
    }

    public TypedTargetBuilder<S, T> when(Predicate<S> condition) {
        this.condition = condition;
        finalizeMappingRule();
        return parent;
    }

    public TypedTargetBuilder<S, T> end() {
        finalizeMappingRule();
        return parent;
    }

    private void finalizeMappingRule() {
        // Use the fully typed mapping rule
        MappingRule rule = new TypedPropertyMapping<>(sourceAccessor, targetSetter, transformation, condition);
        parent.addRule(rule);
    }
}

File: src/main/kotlin/xyz/mahmoudahmed/dsl/typed/TypedTargetPathBuilder.kt
package xyz.mahmoudahmed.dsl.typed

class TypedTargetPathBuilder<S, T, V, U>(
    private val parent: TypedTargetBuilder<S, T>,
    private val sourceAccessor: (S) -> V,
    private val targetSetter: (T, U) -> Unit,
    private var transformation: ((V) -> Any)?,
    private var condition: ((S) -> Boolean)?
) {
    fun using(function: (V) -> U): TypedTargetBuilder<S, T> {
        @Suppress("UNCHECKED_CAST")
        this.transformation = function as (V) -> Any
        finalizeMappingRule()
        return parent
    }

    fun chooseIf(condition: (S) -> Boolean): TypedTargetBuilder<S, T> {
        this.condition = condition
        finalizeMappingRule()
        return parent
    }

    fun end(): TypedTargetBuilder<S, T> {
        finalizeMappingRule()
        return parent
    }

    private fun finalizeMappingRule() {
        val rule = TypedPropertyMapping(sourceAccessor, targetSetter, transformation, condition)
        parent.addRule(rule)
    }
}